MailCore/Utils/URLSchemeHandler.swift

```diff
@@ -19,43 +19,86 @@
 import Foundation
 import WebKit
 
+actor TaskProgressHandler {
+    private var dataTasksInProgress = [String: URLSessionDataTask]()
+    private var urlSchemeTasksInProgress = [String: WKURLSchemeTask]()
+
+    func addDataTask(_ task: URLSessionDataTask, for url: URL) {
+        dataTasksInProgress[url.absoluteString] = task
+    }
+
+    func cancelDataTask(for url: URL) {
+        dataTasksInProgress[url.absoluteString]?.cancel()
+        urlSchemeTasksInProgress[url.absoluteString] = nil
+    }
+
+    func addURLSchemeTask(_ urlSchemeTask: WKURLSchemeTask, for url: URL) {
+        urlSchemeTasksInProgress[url.absoluteString] = urlSchemeTask
+    }
+
+    func didReceive(_ response: URLResponse, for url: URL) {
+        urlSchemeTasksInProgress[url.absoluteString]?.didReceive(response)
+    }
+
+    func didReceive(_ data: Data, for url: URL) {
+        urlSchemeTasksInProgress[url.absoluteString]?.didReceive(data)
+    }
+
+    func didFinish(for url: URL) {
+        urlSchemeTasksInProgress[url.absoluteString]?.didFinish()
+        urlSchemeTasksInProgress[url.absoluteString] = nil
+    }
+
+    func didFailWithError(_ error: Error, for url: URL) {
+        urlSchemeTasksInProgress[url.absoluteString]?.didFailWithError(error)
+        urlSchemeTasksInProgress[url.absoluteString] = nil
+    }
+}
+
 public class URLSchemeHandler: NSObject, WKURLSchemeHandler {
     public static let scheme = "mail-infomaniak"
     public static let domain = "://mail.infomaniak.com"
 
-    private var tasksInProgress = [String: URLSessionDataTask]()
+    private let taskProgressHandler = TaskProgressHandler()
 
     public func webView(_ webView: WKWebView, start urlSchemeTask: WKURLSchemeTask) {
         guard let url = urlSchemeTask.request.url else { return }
-        tasksInProgress[url.absoluteString]?.cancel()
-
-        var components = URLComponents(url: url, resolvingAgainstBaseURL: true)
-        components?.scheme = "https"
-        var request = URLRequest(url: components!.url!)
-        request.addValue(
-            "Bearer \(AccountManager.instance.currentAccount.token.accessToken)",
-            forHTTPHeaderField: "Authorization"
-        )
-        let task = URLSession.shared.dataTask(with: request) { data, response, error in
-            guard (error as? URLError)?.code != .cancelled else { return }
-
-            if let response = response {
-                urlSchemeTask.didReceive(response)
-            }
-            if let data = data {
-                urlSchemeTask.didReceive(data)
-                urlSchemeTask.didFinish()
-            }
-            if let error = error {
-                urlSchemeTask.didFailWithError(error)
+        Task {
+            await taskProgressHandler.cancelDataTask(for: url)
```

Comment:
I may not understand this perfectly, yet if for some reason we "navigate" back and forth you can probably send events multiple times for one single URL.

Reply:
discussed with @PhilippeWeidmann